### PR TITLE
[wasm] wasi: Enable library tests on CI

### LINKED ISSUE #72641: [mono][wasi] Add a CI lane for a new wasi RID
As discussed with @radical offline :)

### PR DESCRIPTION
**Pull request by radical**
- Adds tasks/targets to run a wasi console sample, and tests on CI - for linux, and windows
- Only few tests are enabled by default
- Other tests fail due to GC issues, missing fs/network access etc
- Also, adds `runtime-wasm-optional` pipeline which will have the wasi build with all the tests enabled, and will be RED. This is disabled right now.

Some differences from browser/wasm case:

```
    [wasi] Use the real entrypoint even with async Main

    When the main assembly has a async main, roslyn generates a wrapper
    method which invokes the async main. And in case of wasm that
    async-main method is invoked so it can yield to the host.

    But in case of wasi we invoke the actual entrypoint.
```

Fixes https://github.com/dotnet/runtime/issues/72641 .

### CODE REVIEW

**Comment by pavelsavara** [comment]
I would like to see this merged so that I could start looking at the tests. How can I help ?

**Comment by radical** [comment]
update: I'm cleaning this up now. I have to check the various runtime/native changes/cmake changes to make sure that they are still needed, after the earlier wasi PR being merged. I will open this up for review in a few days.

**Review by pavelsavara — COMMENTED** [review]
good work!

This is large PR and I made many comments, but I don't insist on any of them. 
I would like to merge it soon and fix things incrementally on top of this in further PRs.

File: eng/testing/WasiRunnerTemplate.sh
Comment (radical) [reply eng/testing/WasiRunnerTemplate.sh:1]: Yep. I will add the windows bits after this PR.

File: eng/testing/tests.browser.targets
Comment (radical) [reply eng/testing/tests.browser.targets:1]: There are bits that I'm extracting from this to the base targets file. But this is very much WIP, so tests passing is the check for the changes right now!

File: src/tasks/WasmAppBuilder/wasi/EmitWasmBundleObjectFile.cs
Comment (radical) [reply src/tasks/WasmAppBuilder/wasi/EmitWasmBundleObjectFile.cs:16]: We can remove this, or make it optional in future. Right now, it will unblock this PR.

**Comment by radical** [comment]
I've read the feedback on all the tests targets, and wasi targets files, but ignoring them for now, because they are very much in WIP. Most of them are copies of the existing wasm files, and will get pruned in future PRs.

File: src/libraries/sendtohelix-wasm.targets
Comment (radical) [reply src/libraries/sendtohelix-wasm.targets:41]: This was part of something I was trying. It should be able to drop this now.

**Comment by radical** [comment]
Thank you for the detailed feedback, @pavelsavara! I have updated the PR, and added some more details in the PR description.

File: src/mono/wasi/mono-include/driver.h
Comment (radical) [reply src/mono/wasi/mono-include/driver.h]: This was my original commit message which kinda got lost in re-arranging the commits!
```
    [wasi] Use the real entrypoint even with async Main

    When the main assembly has a async main, roslyn generates a wrapper
    method which invokes the async main. And in case of wasm that
    async-main method is invoked so it can yield to the host.

    But in case of wasi we invoke the actual entrypoint.
```

File: src/mono/mono/utils/mono-threads-wasm.c
Comment (radical) [reply src/mono/mono/utils/mono-threads-wasm.c:64]: This was working earlier, but I kept it as-is for this PR. This and the gc will need fixes.
But this is only affecting wasi, so as long as CI is green, we can merge it in this PR. And make changes in follow up.

**Comment by radical** [comment]
I'm testing enabling tests now windows in https://github.com/dotnet/runtime/pull/81452 .

File: src/mono/wasi/wasi.proj
Comment (pavelsavara) [inline src/mono/wasi/wasi.proj:9]: Could we enable this again before merge or this breaks stuff ? 
cc @mkhamoyan

Comment (radical) [reply src/mono/wasi/wasi.proj:9]: I'll try.

Comment (radical) [reply src/mono/wasi/wasi.proj:9]: With this uncommented, a test build fails with:
```
wasm-ld : error : /workspaces/runtime/artifacts/bin/microsoft.netcore.app.runtime.wasi-wasm/Release/runtimes/wasi-wasm/native/libicudata.a: archive has no index; run ranlib to add one [/workspaces/runtime/src/libraries/System.Buffers/tests/System.Buffers.Tests.csproj]
wasm-ld : error : /workspaces/runtime/artifacts/bin/microsoft.netcore.app.runtime.wasi-wasm/Release/runtimes/wasi-wasm/native/libicui18n.a: archive has no index; run ranlib to add one [/workspaces/runtime/src/libraries/System.Buffers/tests/System.Buffers.Tests.csproj]
wasm-ld : error : /workspaces/runtime/artifacts/bin/microsoft.netcore.app.runtime.wasi-wasm/Release/runtimes/wasi-wasm/native/libicuuc.a: archive has no index; run ranlib to add one [/workspaces/runtime/src/libraries/System.Buffers/tests/System.Buffers.Tests.csproj]
```

Comment (mkhamoyan) [reply src/mono/wasi/wasi.proj:9]: I'l take a look.

Comment (pavelsavara) [reply src/mono/wasi/wasi.proj:9]: https://github.com/dotnet/runtime/issues/81560

File: src/mono/mono/metadata/sgen-mono.c
Comment (pavelsavara) [inline src/mono/mono/metadata/sgen-mono.c:2296]: should be rather `HOST_BROWSER` ?

**Review by pavelsavara — APPROVED** [review]

**Comment by radical** [comment]
Wasi tests are failing on `	g_assert (heap_base - stack_size_pad == data_end);`
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-81052-merge-b56c232ac68c49339e/Microsoft.XmlSerializer.Generator.Tests/1/console.e72d0849.log?helixlogtype=result

**Pull request merged**
